<a href="https://colab.research.google.com/github/strejo4/ColabFiles_UACJ_PADP/blob/main/Practica2_SergioTrejo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Extracción de Información con Expresiones Regulares

**Diseñar un código en Google Colab que permita a los estudiantes cargar el archivo informe_acciones.txt y extraer información clave utilizando expresiones regulares en Python.**

* **Intruccion para cargar informe_acciones.txt directamente a Google Colab.**
* **Muestra un ejemplo de carga de archivos utilizando las herramientas de Colab.**


In [20]:
informe = open('/content/drive/My Drive/ClassFiles/informe_acciones.txt', encoding='latin-1')
for i in range(150): #Imprimir los primeros 10 correos solamente o primeras 150 lineas del archivo de texto
    print(informe.readline().rstrip())

De: mercado@acciones.com
Para: inversor@ejemplo.com
Asunto: Informe de Acción - CSCO

Hola,

Te escribo para informarte sobre el valor actual de la acción de Cisco Systems Inc. (CSCO).
Fecha: 2024-09-02 12:34:26
Valor actual: $446.52

Saludos,
Equipo de Mercado

----------------------------------------

De: mercado@acciones.com
Para: inversor@ejemplo.com
Asunto: Informe de Acción - AMZN

Hola,

Te escribo para informarte sobre el valor actual de la acción de Amazon.com Inc. (AMZN).
Fecha: 2024-09-02 12:34:26
Valor actual: $520.22

Saludos,
Equipo de Mercado

----------------------------------------

De: mercado@acciones.com
Para: inversor@ejemplo.com
Asunto: Informe de Acción - NVDA

Hola,

Te escribo para informarte sobre el valor actual de la acción de NVIDIA Corporation (NVDA).
Fecha: 2024-09-02 12:34:26
Valor actual: $552.66

Saludos,
Equipo de Mercado

----------------------------------------

De: mercado@acciones.com
Para: inversor@ejemplo.com
Asunto: Informe de Acción - MRK

Hol

# Datos importantes:

* El nombre de la accion y su ticker (por ejemplo, Cisco Systems Inc. (CSCO), Amazon.com Inc. (AMZN)).

* La fecha y hora de la medicion de la accion (por ejemplo, Fecha: 2024-09-02 12:34:26).

* El valor actual (por ejemplo, Valor actual: $446.52).

Estos elementos es donde podemos extraer informacion clave ya que nos dan los datos que realmente describen el estado de cada accion en el momento del informe o de la notificacion por correo electronico.

Usa re para extraer la siguiente información:
  * Nombre de la empresa: Busca el patrón del nombre en las líneas del asunto.
  * Fecha del valor de la acción: Encuentra el formato de fecha.
  * Valor de la acción: Extrae números después de "Valor actual:".

Análisis Estadístico:
  * Para cada empresa extraída:
    * Encuentra el valor máximo, mínimo, promedio


In [7]:
# Codigo para extraer los valores de todas las empresas y mostrar estadisticos
import re
# Abrir el archivo
informe = open('/content/drive/My Drive/ClassFiles/informe_acciones.txt', encoding='latin-1')

# Diccionario para guardar datos por empresa
datos_empresas = {}

# Recorrer cada línea del archivo
for linea in informe:
    linea = linea.rstrip()

    # Buscar nombre y ticker de la empresa
    nombre_ticker = re.findall(r'acción de (.+)\s\((.+)\)', linea) # Busca "accion de ", (.+) = nombre de la empresa, \s = espacio, \( y \) = parentesis, (.+) = ticker
    if len(nombre_ticker) > 0:
        nombre, ticker = nombre_ticker[0]
        empresa = f"{nombre} ({ticker})"
        if empresa not in datos_empresas:
            datos_empresas[empresa] = {'fechas': [], 'valores': []}

    # Buscar fecha
    fecha = re.findall(r'^Fecha: (.+)', linea) # Busca "^" inicio de la linea, "Fecha: ", (.+) = el resto de la linea
    if len(fecha) > 0:
        datos_empresas[empresa]['fechas'].append(fecha[0])

    # Buscar valor actual
    valor = re.findall(r'^Valor actual: \$?([0-9.]+)', linea) # Busca "^" inicio de la linea, "Valor actual: ", \$? = signo $ opcional, ([0-9.]+) = numeros y puntos
    if len(valor) > 0:
        datos_empresas[empresa]['valores'].append(float(valor[0]))

# Realizar analisis estadistico e Imprimir tabla con el formato deseado
print(f"{'Empresa':<35}{'Fecha mas reciente':<20}{'Valor Max ($)':<15}{'Valor Min ($)':<15}{'Promedio ($)':<15}'")
for empresa, info in datos_empresas.items():
    if len(info['valores']) > 0:
        fecha_reciente = max(info['fechas'])
        valor_max = max(info['valores'])
        valor_min = min(info['valores'])
        promedio = sum(info['valores']) / len(info['valores'])
        print(f'{empresa:<35}{fecha_reciente:<20}{valor_max:<15.2f}{valor_min:<15.2f}{promedio:<15.2f}')






Empresa                            Fecha mas reciente  Valor Max ($)  Valor Min ($)  Promedio ($)   '
Cisco Systems Inc. (CSCO)          2024-09-02 12:34:26 1499.73        104.72         794.88         
Amazon.com Inc. (AMZN)             2024-09-02 12:34:26 1498.27        103.83         762.75         
NVIDIA Corporation (NVDA)          2024-09-02 12:34:26 1496.06        103.26         806.61         
Merck & Co., Inc. (MRK)            2024-09-02 12:34:26 1495.62        102.43         804.27         
Microsoft Corp. (MSFT)             2024-09-02 12:34:26 1494.97        104.05         819.01         
The Home Depot Inc. (HD)           2024-09-02 12:34:26 1490.53        101.10         798.42         
Salesforce.com Inc. (CRM)          2024-09-02 12:34:26 1497.80        106.33         816.63         
Exxon Mobil Corporation (XOM)      2024-09-02 12:34:26 1498.51        104.62         822.57         
MasterCard Inc. (MA)               2024-09-02 12:34:26 1499.99        102.26         802.1